### DISCUSSION TITLE
DefenderServices check (edge-cases)

### DISCUSSION THREAD

**Discussion by alan-null** [farag2/Sophia-Script-for-Windows#425] [Q&A] [answered]
Didn't know if I should start a discussion or report it as an issue so let me know how do you want to handle code/logic related things in the future (issues or discussion first).

I am currently reading the whole code and found this

https://github.com/farag2/Sophia-Script-for-Windows/blob/6692e83d68e7421c82d90910d52e91db7db6c69a/src/Sophia_Script_for_Windows_10/Module/Sophia.psm1#L300-L307

So there are 3 services that script checks via Foreach
- Windefend
- SecurityHealthService
- wscsvc

When we have following state:
```
Status   Name               DisplayName
------   ----               -----------
Running  Windefend          Microsoft Defender Antivirus Service
Running  SecurityHealthS... Windows Security Service
Running  wscsvc             Security Center
```
everything will work fine. 

The potential problem I see here is when `$Script:DefenderServices` will be set to false by `wscsvc`. Potential edge-case (last service state other than `Running`)
```
Status   Name               DisplayName
------   ----               -----------
Running  Windefend          Microsoft Defender Antivirus Service
Running  SecurityHealthS... Windows Security Service
Stopped  wscsvc             Security Center
```

So basically Sophia uses only **one** flag where state of **3** different services is stored.

Haven't checked how this could actually influence the rest but it looks a bit illogical to me. The same false state will appear when all services are `Stopped` and only last one is `Running`

**Comment by alan-null** [comment]
@farag2 In case you missed. Please take a look 👆

**Comment by farag2** [comment]
I really forgot about this topic. Let take a look...

**Comment by farag2** [comment]
Well, I do not see any problem with the code. And it's not illogical ))

```powershell
@("Windefend", "SecurityHealthService", "wscsvc") | ForEach-Object -Process {
	if ($null -eq (Get-Service -Name $_ -ErrorAction Ignore))
	{
		$Localization.WindowsBroken
		exit
	}
	else
	{
		if ((Get-Service -Name $_).Status -eq "running")
		{
			$Script:DefenderServices = $true
		}
		else
		{
			$Script:DefenderServices = $false
		}
	}
}
```

I check if all services exist. If not, we receive error and exit. Else we check if they're running. Correct me where the logic is broken? I've tested on many machines and VMs to isolate bugs, but logic works)

**Comment by alan-null** [comment]
@farag2 
Here is an example code proving that the current implementation returns different results depending on **order of services**.
First `if` in the code is not an issue I am talking about `else` when all services exist

```powershell
Clear-Host
# get example srvice with stopped status
$exampleStopped = Get-Service | ? { $_.Status -eq "stopped" } | Select-Object -First 1
# get example srvice with running status
$examplerunning = Get-Service | ? { $_.Status -eq "running" } | Select-Object -First 1

# Sophia code
function Test-Implementation {
    param ($services)

    $services | ForEach-Object -Process {
        if ($null -eq (Get-Service -Name $_ -ErrorAction Ignore)) {
            $Localization.WindowsBroken
            exit
        }
        else {
            if ((Get-Service -Name $_).Status -eq "running") {
                $Script:DefenderServices = $true
            }
            else {
                $Script:DefenderServices = $false
            }
        }
    }    
}

# Test scenarios

# services will be different, name doesn't matter but status for this case is STOPPED, RUNNING
$services = @($exampleStopped.Name, $examplerunning.Name)
Write-Host "Services: $($services -join ',')"
Test-Implementation $services
Write-Host "Script:DefenderServices = $($Script:DefenderServices)"
""

# services will be different, name doesn't matter but status for this case is RUNNING, STOPPED
$services = @($examplerunning.Name, $exampleStopped.Name)
Write-Host "Services: $($services -join ',')"
Test-Implementation $services
Write-Host "Script:DefenderServices = $($Script:DefenderServices)"
```

**Result:**
```
Services: aspnet_state,BDESVC
Script:DefenderServices = True

Services: BDESVC,aspnet_state
Script:DefenderServices = False
```

Do you get my point now?

**Comment by farag2** [comment]
Yeah, really, you've convinced me. How to fix?..

**Reply by alan-null** [reply]
I can take a look this week and create PR.

**Comment by farag2** [comment, marked as answer]
#430

**Reply by alan-null** [reply]
Improvement will be implemented here.
